From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: FirebirdTA01 <FirebirdTA01@aol.com>
Date: Sun, 18 Oct 2026 17:01:28 -0500
Subject: [PATCH 15/38] spu: argument hooks take function_arg_info (GCC 10)

function_arg, function_arg_advance, pass_by_reference and
setup_incoming_varargs.  setup_incoming_varargs skips the advance past the
last named argument when there is none (TYPE_NO_NAMED_ARGS_STDARG_P, the
C23 rule the other ports adopted in GCC 13).
---
 gcc/config/spu/spu.cc | 29 ++++++++++++++++-------------
 1 file changed, 16 insertions(+), 13 deletions(-)

diff --git a/gcc/config/spu/spu.cc b/gcc/config/spu/spu.cc
index a29053474bd..8837510591f 100644
--- a/gcc/config/spu/spu.cc
+++ b/gcc/config/spu/spu.cc
@@ -3833,11 +3833,11 @@ spu_function_value (const_tree type, const_tree func ATTRIBUTE_UNUSED)
 }
 
 static rtx
-spu_function_arg (cumulative_args_t cum_v,
-		  machine_mode mode,
-		  const_tree type, bool named ATTRIBUTE_UNUSED)
+spu_function_arg (cumulative_args_t cum_v, const function_arg_info &arg)
 {
   CUMULATIVE_ARGS *cum = get_cumulative_args (cum_v);
+  machine_mode mode = arg.mode;
+  const_tree type = arg.type;
   int byte_size;
 
   if (*cum >= MAX_REGISTER_ARGS)
@@ -3870,10 +3870,12 @@ spu_function_arg (cumulative_args_t cum_v,
 }
 
 static void
-spu_function_arg_advance (cumulative_args_t cum_v, machine_mode mode,
-			  const_tree type, bool named ATTRIBUTE_UNUSED)
+spu_function_arg_advance (cumulative_args_t cum_v,
+			  const function_arg_info &arg)
 {
   CUMULATIVE_ARGS *cum = get_cumulative_args (cum_v);
+  machine_mode mode = arg.mode;
+  const_tree type = arg.type;
 
   *cum += (type && TREE_CODE (TYPE_SIZE (type)) != INTEGER_CST
 	   ? 1
@@ -3906,11 +3908,9 @@ spu_function_arg_padding (machine_mode, const_tree)
 
 /* Variable sized types are passed by reference.  */
 static bool
-spu_pass_by_reference (cumulative_args_t cum ATTRIBUTE_UNUSED,
-		       machine_mode mode ATTRIBUTE_UNUSED,
-		       const_tree type, bool named ATTRIBUTE_UNUSED)
+spu_pass_by_reference (cumulative_args_t, const function_arg_info &arg)
 {
-  return type && TREE_CODE (TYPE_SIZE (type)) != INTEGER_CST;
+  return arg.type && TREE_CODE (TYPE_SIZE (arg.type)) != INTEGER_CST;
 }
 
 
@@ -4095,8 +4095,9 @@ spu_gimplify_va_arg_expr (tree valist, tree type, gimple_seq * pre_p,
    in the stack then save no registers.  Set pretend_args_size to the
    amount of space needed to save the registers. */
 static void
-spu_setup_incoming_varargs (cumulative_args_t cum, machine_mode mode,
-			    tree type, int *pretend_size, int no_rtl)
+spu_setup_incoming_varargs (cumulative_args_t cum,
+			    const function_arg_info &arg,
+			    int *pretend_size, int no_rtl)
 {
   if (!no_rtl)
     {
@@ -4106,8 +4107,10 @@ spu_setup_incoming_varargs (cumulative_args_t cum, machine_mode mode,
       int ncum = *get_cumulative_args (cum);
 
       /* cum currently points to the last named argument, we want to
-         start at the next argument. */
-      spu_function_arg_advance (pack_cumulative_args (&ncum), mode, type, true);
+         start at the next argument.  A C23 function with no named
+         argument has none to step over (GCC 13).  */
+      if (!TYPE_NO_NAMED_ARGS_STDARG_P (TREE_TYPE (current_function_decl)))
+	spu_function_arg_advance (pack_cumulative_args (&ncum), arg);
 
       offset = -STACK_POINTER_OFFSET;
       for (regno = ncum; regno < MAX_REGISTER_ARGS; regno++)
